### PR TITLE
Circular dependency issue fix, RequireJS support, and other fixes.

### PR DESCRIPTION
**Pull request by circlingthesun**
This pull request contains support for RequireJS and fixes some issues.

The first issue I found was that the key used in the cache was inconsistent. In some places the file path in the gulp file was used, and in other places the path reference as found in a file which may be absolute or relative. This resulted in cache misses and duplicate hash entries. I changed cache lookups to always use the absolute path.

Secondly the recursive dependency lookup stopped when it found a circular reference. In the case of a.html -> b.html -> a.html, when computing the hash for a.html it will try compute the hash for b.html but then recursively find a.html again. The cycle will be detected and a will not be hashed again. b.html will be hashed and it will be included in the hash for a.html. However, b.html contains a reference to a.html. When this reference is replaced, it modifies the content of b.html, so the hash of b.html should change, but does not. When you next deploy your site the old cached version of b.html will be found that links to the old version of a.html and result in a 404. I solved this by keeping a list of files in the cycle. Once a.html is done hashing, the hash of a.html will be back propagated to all the files in the chain.

The fix for circular dependencies caused some breakages in the unit tests. The reason for this is that when circular references are present, the order in which you hash files affects the hash value. So hashing a.html first will lead to a different result than when you hash b.html first. For this reason I export the cache from the tool so that hashes can be looked up instead of recomputing individual file hashes which may not be consistent across runs. I refactored all the test cases to not recompute hashes but look them up.

A missing feature that bugged me was setting the base path of references in your file tree. I've added support for multiple base paths via options.base. I also added a feature that lets you write your an aggregate hash of all files. I use this to check for updates from my app. There are also other small style fixes.

I hope you like.

### CODE REVIEW

**Comment by circlingthesun** [comment]
I just tried version 0.8.17. There is major breakage on my current project. Am I correct in saying that this PR has pretty much been undone?

**Comment by smysnk** [comment]
The code for dependency resolution was way too complex and needed to be refactored as to allow for some form of maintainability.  Originally I set out to clean up your PR but found it too difficult to follow and lacking unit tests.  Sorry but I n the end I decided a rewrite was in store. 

As part of the rewrite I decided that nested dependency resolution was over kill and have settle instead on basic 1-level deep resolution. I am not sure of any specific scenarios that more than 1 level references should affect a hash.

In regards to your issues with the latest, is it incorrect hash or references are not being resolved?  Please describe the scenario so tests can be created around it.

**Comment by smysnk** [comment]
Also in regards to the PR, another change happened that was suppose to help reduce complexity.  That is possible references are no resolved from the file contents, instead it has been reversed.  

Given the current file under going revisioning, all possible permutations of relative/absolute reference representations are tried. 

This should help in speed of resolution as it doesn't need to check existence of files on all possible regex matches as per the old method.  Also should help resolve amd/commonjs replacements without obscure regular expressions.

**Comment by circlingthesun** [comment]
Thanks for clarifying. Yeah, the codebase was a bit hard to follow. The current implementation is a bit too simple for my needs though.

The new version firstly breaks as it appears to no longer work with AMD (at least for my setup). The one instance of when 1 level references are not sufficient is circular dependencies.

Consider the case where you have index.html that references about.html and about.html has a link to index.html. If you resolve references one level deep you might end up with index.123.html and about.567.html. On the surface this looks fine. Now you set the cache expiration to 10 years and the files are stuck in the browser cache of whoever visited your site. Now if you update index.html you end up with index.891.html and about.567.html remains unchanged. A previous visitor that now revisits your site and visits index.891.html. So far so good. The visitor then clicks through to about.567.html whom he or she had visited previously. The visitor now gets the cached version of about.567.html which has a reference to index.123.html instead of index.891.html. If the link back to index.html is clicked the user accesses the old index.123.html instead of index.891.html.

I discovered this problem the hard way in production and it took me more than a week to discover the root cause (after much client upset).

The way I fixed this in my PR was to keep track of the stack when resolving references recursively. When revisioning index.html I would find about.html and recursively resolve the references in about.html. At this point the stack would be ['index.html, 'about.html']. I would then find a reference to index.html in about.html. At this point the stack would be ['index.html', 'about.html', 'index.html']. At this point a circular reference is detected. The recursion stops at this point and the list of files in the cycle is saved. When the stack unrolls and hash for index.html is determined, the hash for files in the cycle needs to be updated. In this case its only about.html that needs an updated hash because it references index.html and the hash needs to reflect any changes in index.html. If the the cycle was longer, the file that reference about.html would be affected by the updated hash of about.html. For this reason all the files in the cycles needs to have the hash of index.html appended because of this ripple effect.

I'm not sure how you would do this in the new setup.

**Comment by circlingthesun** [comment]
So without the circular dependency detection and handling my hashes will most certainly be wrong. The more obviously issue was incorrect reference resolution (probably) and gibberish output.

Input:
```
define([
    'angular',
    'app/app',
    ], function (angular, app) {
            'use strict';
...
```

Output:
```
define(notifications.c9feab60.html[
notifications.c9feab60.html  notifications.c9feab60.html  notifications.c9feab60.html'angular'notifications.c9feab60.html,
notifications.c9feab60.html  notifications.c9feab60.html  notifications.c9feab60.html'app/app'notifications.c9feab60.html,
notifications.c9feab60.html  notifications.c9feab60.html  notifications.c9feab60.html],notifications.c9feab60.html function notifications.c9feab60.html(angular,notifications.c9feab60.html app)notifications.c9feab60.html {notifications.c9feab60.html
 notifications.c9feab60.html  notifications.c9feab60.html  notifications.c9feab60.html  notifications.c9feab60.html  notifications.c9feab60.html  notifications.c9feab60.html 'use strict'notifications.c9feab60.html;
notifications.c9feab60.html
...
```

**Comment by smysnk** [comment]
I've added a test here that I believe recreates the scenario described, and it passes.
https://github.com/smysnk/gulp-rev-all/blob/master/test.js#L73

This was a problem with the old method of reference resolution because of the way it looked up files, the new method does not suffer from this issue.  The reason for this is all 1st level references are resolved first and then after all is done, it will calculate the final hash. 

There does however seem to be a genuine issue with with that define statement, very odd.

**Comment by circlingthesun** [comment]
Interesting. I'll try to find some time to understand the new code and see if I can make things work again. The codebase looks much cleaner so it will be great if I can. Thanks for shedding some light on this.

**Comment by smysnk** [comment]
It appears to be matching `notifications.c9feab60.html` to a white space or tab character for replacement.  The file of the contents you have posted, where is it in reference to notifications.html?  Also it appears like `app/app` reference is not getting resolved, where is this file located in reference to root?

There is one option that might need to be added to resolve those references, that is specifying multiple roots.

**Comment by circlingthesun** [comment]
I've updated the unit test you wrote at https://github.com/circlingthesun/gulp-rev-all/tree/circular-test.  There is one more link in the chain and now it fails.

The circular reference is main.html -> about.html -> gps.html -> main.html. When gps.html is updated the hash for about.html should be updated because it references gps.html. Because main.html references about.html, and the hash for about.html has changed, it too needs a new hash because the changed link inside changed the content.

**Comment by smysnk** [comment]
For some reason I thought that beyond 1 reference resolution wasn't needed, but in case of html files, you're certainly right.  I've added back recursive reference resolution in v0.8.18, test passes now.
